File: sdk/python/src/graft/modules/dht.py
"""GRAFT DHT Module"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from graft.device import GraftDevice

from graft.core.constants import CMD


@dataclass
class DHTReading:
    temperature: float
    humidity: float


class GraftDHT:
    def __init__(self, device: GraftDevice, module_id: int = 0) -> None:
        self._device = device
        self._mod_id = module_id

    async def begin(self, pin: int, sensor_type: int = 22) -> None:
        payload = bytes([self._mod_id, 0x01, pin, sensor_type])
        await self._device._send_command(CMD.MOD_CMD, payload)

    async def read(self) -> DHTReading:
        payload = bytes([self._mod_id, 0x02])
        resp = await self._device._send_command(CMD.MOD_CMD, payload)
        if len(resp) >= 9:  # module_id(1) + 2 floats(8)
            temp, hum = struct.unpack_from("<ff", resp, 1)
            return DHTReading(temperature=temp, humidity=hum)
        return DHTReading(temperature=0.0, humidity=0.0)

File: sdk/js/src/modules/dht.ts
/**
 * GRAFT DHT Module — Typed JS wrapper
 */

import type { GraftDevice } from '../device.js'

export interface DHTReading {
  temperature: number
  humidity: number
}

export class GraftDHT {
  private _device: ReturnType<GraftDevice['module']>

  constructor(device: GraftDevice) {
    this._device = device.module('dht')
  }

  /** Initialize DHT sensor. type: 11 for DHT11, 22 for DHT22. */
  async begin(pin: number, type: 11 | 22 = 22): Promise<void> {
    await this._device.cmd(0x01, new Uint8Array([pin, type]))
  }

  /** Trigger a reading. Returns temperature and humidity. */
  async read(): Promise<DHTReading> {
    const resp = await this._device.cmd(0x02)
    if (resp.length >= 8) {
      const view = new DataView(resp.buffer, resp.byteOffset, resp.byteLength)
      return {
        temperature: view.getFloat32(0, true),
        humidity: view.getFloat32(4, true),
      }
    }
    return { temperature: 0, humidity: 0 }
  }
}
